neps/nep-0393.md

```diff
@@ -0,0 +1,552 @@
+/// a wallet. See also `Burn` event.
+/// Must be emitted by an SBT registry.
+type Revoke {
+  ctr: AccountId;  // SBT Contract revoking the tokens
+  tokens: u64[];  // list of token ids.
+}
+
+/// An event emitted when existing tokens are burned and removed from the registry.
+/// Must be emitted by an SBT registry.
+type Burn {
+  ctr: AccountId;  // SBT Contract burning the tokens
+  tokens: u64[];  // list of token ids.
+}
+
+/// An event emitted when an account is banned within the emitting registry.
+/// Registry must add the `account` to a list of accounts that are not allowed to get any SBT
+/// in the future.
+/// Must be emitted by an SBT registry.
+type Ban = AccountId;
+
+/// An event emitted when soul transfer is happening: all SBTs owned by `from` are transferred
+/// to `to`, and the `from` account is banned (can't receive any new SBT).
+/// Must be emitted by an SBT registry.
+/// Registry MUST also emit `Ban` whenever the soul transfer happens.
+type SoulTransfer {
+  from: AccountId;
+  to: AccountId;
+}
+```
+
+Whenever a recovery is made in a way that an existing SBT is burned, the `Burn` event MUST be emitted. If `Revoke` burns token then `Burn` event MUST be emitted instead of `Revoke`.
+
+### Example SBT Contract functions
+
+Although the transaction functions below are not part of the SBT smart contract standard (depending on a use case, they may have different parameters), we present here an example interface for SBT issuance and we also provide a reference implementation.
+These functions must relay calls to an SBT registry, which will emit appropriate events.
+We recommend that all functions related to an event will take an optional `memo: Option<String>` argument for accounting purposes.
+
+```rust
+trait SBT {
+    /// the function should overwrite `metadata.class = class`.
+    /// Must provide enough NEAR to cover registry storage cost.
+    // #[payable]
+    fn sbt_mint(
+        &mut self,
+        account: AccountId,
+        class: u64,
+        metadata: TokenMetadata,
+        memo: Option<String>,
+    ) -> TokenId;
+
+    /// Creates a new, unique token and assigns it to the `receiver`.
+    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
+    /// Must provide enough NEAR to cover registry storage cost.
+    // #[payable]
+    fn sbt_mint_multi(
+        &mut self,
+        token_spec: Vec<(AccountId, TokenMetadata)>,
+        memo: Option<String>,
+    ) -> Vec<TokenId>;
+
+    // #[payable]
+    fn sbt_recover(&mut self, from: AccountId, to: AccountId, memo: Option<String>);
+
+    fn sbt_renew(&mut self, tokens: Vec<TokenId>, expires_at: u64, memo: Option<String>);
+
+    fn sbt_revoke(token: Vec<u64>, memo: Option<String>) -> bool;
+}
+```
+
+## Reference Implementation
+
+- Common [type definitions](https://github.com/alpha-fi/i-am-human/tree/master/contracts/sbt) (events, traits).
+- https://github.com/alpha-fi/i-am-human/tree/master/contracts/soulbound
+
+## Example Flows
```

Comment:
@robert-zaremba I would argue that this section (diagrams) is so important for understanding of the NEP that it should be at the top of the NEP right in the Specification section.

Reply:
@robert-zaremba It is hard for me to keep small bits in memory unless I visualize / structure them, so diagrams are really helpful. At the very least, I would reference them in the first section(s).